cmd/internal/find/types/entrySchema.go

```diff
@@ -0,0 +1,188 @@
+			panic("gatherNodes called with an empty type ID!")
+		}
+		node := nodes[s.TypeID]
+		if node == nil {
+			node = &EntrySchema{
+				EntrySchema: s,
+				Children:    make(map[string]*EntrySchema),
+			}
+			nodes[s.TypeID] = node
+		}
+		if len(node.Children) > 0 {
+			// We already gathered everything we need to know about this node
+			return
+		}
+		for _, child := range s.Children {
+			// The node's children will be filled in after all the nodes have been
+			// gathered.
+			node.Children[child.TypeID] = nil
+			gatherNodes(child)
+		}
+	}
+	gatherNodes(s)
+
+	for _, node := range nodes {
+		for childTypeID := range node.Children {
+			node.Children[childTypeID] = nodes[childTypeID]
+		}
+	}
+	return nodes[s.TypeID]
+}
+
+// Prune prunes s to contain only nodes that satisfy p. It modifies s'
+// state so it is not an idempotent operation.
+func Prune(s *EntrySchema, p EntrySchemaPredicate) *EntrySchema {
+	keep := evaluateNodesToKeep(s, p)
+	if !keep[s.TypeID] {
+		return nil
+	}
+
+	visited := make(map[string]bool)
+	var prune func(s *EntrySchema)
+	prune = func(s *EntrySchema) {
+		if visited[s.TypeID] {
+			return
+		}
+		visited[s.TypeID] = true
+		for _, child := range s.Children {
+			if !keep[child.TypeID] {
+				delete(s.Children, child.TypeID)
+			} else {
+				prune(child)
+			}
+		}
+	}
+	prune(s)
+
+	return s
+}
+
+/*
+"result" represents the returned value, which is a map of <node> => <keep?>.
+Thus, result[N] == false means we prune N, result[N] == true means we keep
+it. result[N] == true if:
+	* p(N) returns true
+	* p(N) returns true for at least one of N's children (ignoring self-loops)
+Since entry schemas can contain cycles (e.g. the volume.dir class in the
+volume package), and since they do not strictly adhere to a tree structure
+(e.g. paths A-B-C and A-D-C are possible), we cannot completely calculate
+"result" in a single iteration. For example in the A-B-C, A-D-C case, assume
+that p returns false for A, B, C and D. Then we keep B and D if we keep C,
+and we keep C if we keep any of its children. If C does not contain a cycle,
+then we can recurse back the information to B and D. If C does contain a cycle,
+say back to B, then it is difficult, if not impossible, to coordinate all the
+updates in a single iteration. However if we let "result[N] == p(N) for all N
+such that either p(N) == true or N has no children" be our starting state, then
+we can iteratively evaluate result[V] for all other nodes V using the information
+provided by a previous iteration. The terminating condition is when an iteration
+does not update "result". This is possible if "result" contains all nodes N, or if
+there are some indeterminate nodes V. In the latter case, every node M in V is either
+part of a cycle, or there is more than one way to get to M from the root. In either
+case, we can say that result[M] == false for those nodes.
+
+NOTE: To see why result[M] == false for all M in V, we do a proof by contradiction.
```

Comment:
Self-loops (like `volume.Dir` => `volume.Dir`) will likely be the most common cycles.